### ISSUE TITLE
Imputación_RandomForest - Yefer Andersson Mamani Chambi

### ISSUE THREAD

**Issue by YeferAndersson** [frdtorres/mainf24-2-BIDA#4]
import pandas as pd
import numpy as np
import requests
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

def fetch_data_from_api(url):
    """Obtiene los datos de la API y los convierte en un DataFrame."""
    response = requests.get(url)
    if response.status_code == 200:
        print("Datos obtenidos exitosamente de la API.")
        return pd.DataFrame(response.json())
    else:
        raise ValueError(f"Error al obtener los datos: {response.status_code}")

def preprocess_data(df):
    """Preprocesa los datos y codifica columnas categóricas."""
    df_encoded = df.copy()
    # Convertir columnas categóricas en variables numéricas
    for col in ['item_name', 'item_type', 'received_by', 'time_of_sale']:
        df_encoded[col] = df_encoded[col].astype('category').cat.codes
    return df_encoded

def split_data_for_imputation(df_encoded):
    """Separa los datos en aquellos con y sin valores faltantes en 'transaction_type'."""
    df_train = df_encoded[df_encoded['transaction_type'].notnull()].copy()
    df_missing = df_encoded[df_encoded['transaction_type'].isnull()].copy()

    # Codificar 'transaction_type' para el modelo
    df_train['transaction_type'] = df_train['transaction_type'].astype('category').cat.codes
    return df_train, df_missing

def train_random_forest_model(X_train, y_train):
    """Entrena un modelo de Random Forest."""
    rf = RandomForestClassifier(n_estimators=100, random_state=42)
    rf.fit(X_train, y_train)
    return rf

def evaluate_model(rf, X_test, y_test):
    """Evalúa el modelo y muestra la exactitud."""
    y_pred = rf.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    

def impute_missing_values(rf, df_missing):
    """Predice los valores faltantes en 'transaction_type'."""
    X_missing = df_missing.drop(['transaction_type', 'order_id', 'date'], axis=1)
    df_missing.loc[:, 'transaction_type'] = rf.predict(X_missing)
    return df_missing

def main():
    # URL de la API
    url = "http://127.0.0.1:5001/sales"

    # 1. Obtener los datos de la API
    df = fetch_data_from_api(url)

    # 2. Mostrar valores faltantes por columna
    print("\nValores faltantes por columna:")
    print(df.isnull().sum())

    # 3. Preprocesar los datos (sin incluir 'date' en el modelo)
    df_encoded = preprocess_data(df)

    # 4. Separar los datos para la imputación
    df_train, df_missing = split_data_for_imputation(df_encoded)

    # 5. Preparar los datos para entrenamiento
    X_train, X_test, y_train, y_test = train_test_split(
        df_train.drop(['transaction_type', 'order_id', 'date'], axis=1),
        df_train['transaction_type'],
        test_size=0.2,
        random_state=42
    )

    # 6. Entrenar el modelo Random Forest
    rf = train_random_forest_model(X_train, y_train)

    # 7. Evaluar el modelo
    evaluate_model(rf, X_test, y_test)

    # 8. Imputar los valores faltantes
    df_missing_imputed = impute_missing_values(rf, df_missing)

    # 9. Reconstruir el DataFrame completo
    df_imputed = pd.concat([df_train, df_missing_imputed], axis=0)

    # 10. Decodificar 'transaction_type' al valor original
    transaction_type_mapping = {
        idx: category for idx, category in enumerate(df['transaction_type'].astype('category').cat.categories)
    }
    df_imputed['transaction_type'] = df_imputed['transaction_type'].map(transaction_type_mapping)

    # 11. Verificar valores faltantes después de la imputación
    print("\nValores faltantes después de la imputación:")
    print(df_imputed.isnull().sum())

    # 12. Mostrar solo las filas donde se imputaron datos
    print("\nFilas con datos imputados:")
    print(df_missing_imputed)

if __name__ == "__main__":
    main()
